File: lib/libesp32/Berry/tools/coc/macro_table.cpp
/********************************************************************
** Copyright (c) 2018-2020 Guan Wenliang
** This file is part of the Berry default interpreter.
** skiars@qq.com, https://github.com/Skiars/berry
** See Copyright Notice in the LICENSE file or at
** https://github.com/Skiars/berry/blob/master/LICENSE
********************************************************************/
#include "macro_table.h"
#include <regex>
#include <fstream>
#include <sstream>

std::string macro_table::readfile(const std::string &filename)
{
    std::ifstream in(filename);
    std::ostringstream tmp;
    tmp << in.rdbuf();
    return tmp.str();
}

int macro_table::parse_value(std::string str)
{
    if (!str.length()) {
        return 1; /* defined a macro name but no content, considered true */
    }
    if (!(str[0] >= '0' && str[0] <= '9')) {
        return 1;
    }
    return atoi(str.c_str());
}

void macro_table::scan_file(const std::string &filename)
{
    std::string str(readfile(filename));
    std::regex reg("(?:\\n|$)\\s*#define\\s+(\\w+)[ \\t]+(\\w+)");
    std::sregex_iterator it(str.begin(), str.end(), reg);
    std::sregex_iterator end;
    while (it != end) {
        m_map[it->str(1)] = parse_value(it->str(2));
        ++it;
    }
}

bool macro_table::query(const std::string &str) const
{
    std::regex reg("(!?)(\\w+)");
    std::match_results<std::string::const_iterator> res;
    if (regex_match(str, res, reg)) {
        auto it = m_map.find(res[2]);
        int value = it == m_map.end() ? 0 : it->second;
        return res[1] == "!" ? value == 0 : value != 0;
    }
    return 0;
}

std::map<std::string, int> macro_table::table() const
{
    return m_map;
}

File: lib/libesp32/berry/tools/coc/macro_table.py
import re

def int_safe(v):
    try:     return int(v, 0)
    except:  return 0

class macro_table:
    pat = re.compile("(?:\\n|$)\\s*#define\\s+(\\w+)[ \\t]+(\\w*)", re.MULTILINE)
    pat_query = re.compile("(!?)(\\w+)")

    def __init__(self):
        self.map = {}
    
    # def readfile(self, filename):
    #     with open(filename) as f:
    #         return f.read()

    def parse_value(self, s):
        if len(s) == 0:             return 1    # defined a macro name but no content, considered true
        if not s[0].isnumeric():    return 1
        return int_safe(s)
    
    def scan_file(self, filename):
        str = ""
        with open(filename, encoding='utf-8') as f:
           str = f.read()
        r = macro_table.pat.findall(str)
        for it in r:
            # print(f"> it0:{it[0]} it1:{it[1]}")
            self.map[it[0]] = self.parse_value(it[1])
    
    def query(self, s):
        r = macro_table.pat_query.search(s)
        value = False
        if r:
            bang = r[1]
            name = r[2]
            # print(f">query: bang={bang} name={name} value={self.map.get(name)}")
            if name in self.map:
                value = int(self.map[name]) != 0
            if bang == "!":
                value = not value
        # print(f">query: {s}:{value}")
        return value

if __name__ == '__main__':
    # from hash_map import *
    pat = re.compile("(?:\\n|$)\\s*#define\\s+(\\w+)[ \\t]+(\\w+)", re.MULTILINE)
    s = "aaa\n#define A 1  // a \n  #define B 2\n#define C \n#define D 0 \n  #define E     11 \na"
    